From 24095aa347a32673cf220fc0bd0b78d28ba0a69e Mon Sep 17 00:00:00 2001
From: John Fastabend <john.r.fastabend@intel.com>
Date: Wed, 23 Feb 2011 05:58:03 +0000
Subject: [PATCH] ixgbe: enable ndo_tc_setup
Patch-mainline: v2.6.39-rc1
Git-commit: 24095aa347a32673cf220fc0bd0b78d28ba0a69e
References: bnc#687049, fate#311821

This patch adds the ndo_tc_setup to ixgbe. By default we set
the device to use strict priority.

Signed-off-by: John Fastabend <john.r.fastabend@intel.com>
Tested-by: Ross Brattain <ross.b.brattain.@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Joey Lee <jlee@suse.com>

---
 drivers/net/ixgbe/ixgbe_main.c |   24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

--- a/drivers/net/ixgbe/ixgbe_main.c
+++ b/drivers/net/ixgbe/ixgbe_main.c
@@ -4172,7 +4172,10 @@ int ixgbe_setup_tc(struct net_device *de
 	if (!tc) {
 		netdev_reset_tc(dev);
 	} else {
-		if (netdev_set_num_tc(dev, tc))
+		struct ixgbe_adapter *adapter = netdev_priv(dev);
+
+		/* Hardware supports up to 8 traffic classes */
+		if (tc > MAX_TRAFFIC_CLASS || netdev_set_num_tc(dev, tc))
 			return -EINVAL;
 
 		/* Partition Tx queues evenly amongst traffic classes */
@@ -4182,6 +4185,22 @@ int ixgbe_setup_tc(struct net_device *de
 			netdev_set_tc_queue(dev, i, q, offset);
 			offset += q;
 		}
+
+		/* This enables multiple traffic class support in the hardware
+		 * which defaults to strict priority transmission by default.
+		 * If traffic classes are already enabled perhaps through DCB
+		 * code path then existing configuration will be used.
+		 */
+		if (!(adapter->flags & IXGBE_FLAG_DCB_ENABLED) &&
+		    dev->dcbnl_ops && dev->dcbnl_ops->setdcbx) {
+			struct ieee_ets ets = {
+					.prio_tc = {0, 1, 2, 3, 4, 5, 6, 7},
+					      };
+			u8 mode = DCB_CAP_DCBX_HOST | DCB_CAP_DCBX_VER_IEEE;
+
+			dev->dcbnl_ops->setdcbx(dev, mode);
+			dev->dcbnl_ops->ieee_setets(dev, &ets);
+		}
 	}
 	return 0;
 }
@@ -6424,6 +6443,9 @@ static const struct net_device_ops ixgbe
 	.ndo_vlan_rx_add_vid	= ixgbe_vlan_rx_add_vid,
 	.ndo_vlan_rx_kill_vid	= ixgbe_vlan_rx_kill_vid,
 	.ndo_do_ioctl		= ixgbe_ioctl,
+#ifdef CONFIG_IXGBE_DCB
+	.ndo_setup_tc           = ixgbe_setup_tc,
+#endif
 #ifdef CONFIG_NET_POLL_CONTROLLER
 	.ndo_poll_controller	= ixgbe_netpoll,
 #endif
